File: credit/credit.c
#include <cs50.h>
#include <stdio.h>
#include <math.h>
#include <stdlib.h>

string checkCard(long cardNb);
bool hasValidChecksum(long cardNum);
int getNbLength(long number);
int getFirstDigits(long number, int digits);

int main(void)
{
    // Get input
    long cardNb = get_long("Credit card number (eg. 4242424242424242): ");
    // Check card
    string result = checkCard(cardNb);
    // Print result
    printf("%s", result);
}

/**
 * Check wether the card is a valid Visa, Amex or Mastercard credit card
 */
string checkCard(long cardNb)
{
    int length = getNbLength(cardNb);
    int firstDigit = getFirstDigits(cardNb, 1);
    int first2Digits = getFirstDigits(cardNb, 2);

    // Check if has valid checksum, then check the card's type
    if (hasValidChecksum(cardNb))
    {
        // Visa: 13 or 16 digits and starts with 4
        if ((length == 13 || length == 16) && firstDigit == 4)
        {
            return "VISA\n";
        }
        // Amex: 15 digits and starts with 34 or 37
        else if (length == 15 && (first2Digits == 34 || first2Digits == 37))
        {
            return "AMEX\n";
        }
        // Mastercard: 16-digits and starts with 51, 52, 53, 54, or 55
        else if (length == 16 && (first2Digits >= 51 && first2Digits <= 55))
        {
            return "MASTERCARD\n";
        }
    }
    return "INVALID\n";
}

/**
 * Check wether the card's number is compatible with the Luhn’s Algorithm
 */
bool hasValidChecksum(long cardNum)
{
    // 1. Multiply "1 over 2 digits" by 2, starting with the number’s second-to-last digit. Then add those products’ digits together.
    long tempCardNum1 = cardNum / 10;
    int sum1 = 0;
    while (tempCardNum1 > 0)
    {
        int lastDigit = tempCardNum1 % 10; // modulo 10 = get last digit
        int doubled = lastDigit * 2; // multiply by 2
        // extract the products' digits and add them
        if (doubled < 10)
        {
            sum1 = sum1 + doubled;
        }
        else
        {
            sum1 = sum1 + (doubled % 10) + (floor(doubled / 10));
        }
        tempCardNum1 = tempCardNum1 / 100; // remove 2 last digits before looping
    }
    // 2. "Add the sum to the sum of the digits that weren’t multiplied by 2."
    long tempCardNum2 = cardNum;
    int sum2 = 0;
    while (tempCardNum2 > 0)
    {
        int lastDigit = tempCardNum2 % 10;
        sum2 = sum2 + lastDigit;
        tempCardNum2 = tempCardNum2 / 100;
    }
    // 3. If the total’s last digit is 0, the number is valid!
    if ((sum1 + sum2) % 10 == 0)
    {
        return true;
    }
    return false;
}

int getNbLength(long number)
{
    return (number == 0) ? 1 : log10(number) + 1;
}

int getFirstDigits(long number, int digits)
{
    int length = getNbLength(number);
    return number / pow(10, length - digits);
}

File: sentimental-credit/credit.py
def main():
    # Prompt user until the input is a numeric value
    while True:
        card_nb = input("Number: ")
        if not card_nb.isnumeric():
            continue
        # Check card validity and print its type
        if check_luhn_algorythm(card_nb):
            print_type(card_nb)
        else:
            print("INVALID")
        break


def check_luhn_algorythm(card_nb):
    """Check that the input is a valid Credit Card number, based on the Luhn's algorythm. Return True if is a valid number, False otherwise"""
    # Base case: Check length
    if len(card_nb) not in [13, 15, 16]:
        return False

    sum = 0
    # Multiply every other digit by 2, starting with the number’s second-to-last digit, and then add those products’ digits together.
    # `[::-1]` reverses the string and `[1::2]` returns every-two digits, starting from the 2nd digit (index 1)
    for d in card_nb[::-1][1::2]:
        for i in str(int(d)*2):
            sum += int(i)
    # Add the sum to the sum of the digits that weren’t multiplied by 2.
    # `[::2]` returns every-two digits, starting from the 1st digit (index 0)
    for d in card_nb[::-1][::2]:
        sum += int(d)
    # If the total’s last digit is 0 (or, put more formally, if the total modulo 10 is congruent to 0), the number is valid!
    if str(sum)[1] != "0":
        return False

    return True


def print_type(card_nb):
    """Print the type of the credit card: AMEX, VISA or MASTERCARD"""
    # American Express uses 15-digit numbers and start with 34 or 37
    if len(card_nb) == 15 and int(card_nb[:2]) in [34, 37]:
        print("AMEX")
    # MasterCard uses 16-digit numbers and start with 51, 52, 53, 54, or 55
    elif len(card_nb) == 16 and int(card_nb[:2]) in [51, 52, 53, 54, 55]:
        print("MASTERCARD")
    # Visa uses 13- and 16-digit numbers and start with 4
    elif len(card_nb) in [13, 16] and card_nb[:1] == "4":
        print("VISA")
    else:
        print("INVALID")
    return


if __name__ == "__main__":
    main()
